prov/util/src/util_attr.c

```diff
@@ -721,11 +722,52 @@ int ofi_check_tx_attr(const struct fi_provider *prov,
 	return 0;
 }
 
-int ofi_check_info(const struct util_prov *util_prov, uint32_t api_version,
-		   const struct fi_info *user_info)
+/* if there are multiple fi_info in the provider */
+int ofi_check_info_ex(const struct util_prov *util_prov, uint32_t api_version,
+		      const struct fi_info *user_info,
+		      struct fi_info **ret_info)
 {
 	const struct fi_info *prov_info = util_prov->info;
 	const struct fi_provider *prov = util_prov->prov;
+	struct fi_info *info = NULL;
+	int ret, first = 1, res = -FI_ENODATA;
+
+	for ( ; prov_info; prov_info = prov_info->next) {
+		ret = ofi_check_info(util_prov, prov_info,
+				     api_version, user_info);
+	    	if (ret)
+			continue;
+		if (ret_info) {
+			info = fi_dupinfo(prov_info);
+			if (first) {
+				if (!info)
+					goto first;
+				*ret_info = info;
+				first = 0;
+			}
+			if (!info)
+				goto other;
+			info = info->next;
+		}
+		/* found one valid fi_info at least */
+		res = FI_SUCCESS;
+	}
+
+	return res;
+other:
+	fi_freeinfo(*ret_info);
+first:
+	FI_INFO(prov, FI_LOG_CORE,
+		"cannot copy info\n");
+	return -FI_ENOMEM;
+}
```

Comment:
Thanks, I got your point of view. It sounds reasonable and I'm in agreement with you that the proposed approach would be cleaner. 
But something stays unclear for me:
If we would like to drop duplication of info here, where the successfully verified info should duplicated?

Just to summarize:
1. ofi_check_info() is already doing that.
2. ofi_prov_check_info() will be added to verify a single user_info against a provider info list. (But what about duplication, see my question above.)
3. util_getinfo() will be updated. The ND provider already uses it, I'll update the verbs provider if possible.